### DISCUSSION TITLE
Stuck with customizing firmware to change calculation method

### DISCUSSION THREAD

**Discussion by FabsFabios** [klaasnicolaas/home-assistant-glow#409] [Q&A] [answered]
Hi,

I am feeling defeated after having spent an entire day and not gotten any further in my endeavor to add a simple line of code because I don't understand how to customize the firmware. I am trying to follow the guide in the FAQ about abrupt consumption with solar and tried to implement `method: Left` , but I have no idea how to. 

I installed Home Assistant Glow and it shows as adopted in ESPHome. When I edit my home-assistant-glow-28e9a8.yaml packages (3 YAML files) are listed and refer to Github. One of those contain the pulse_meter.yaml file where I need to add `method: left`. I downloaded the file and added the line. However, how am I going about compiling the firmware from ESPHome with the modified pulse_meter.yaml file when my home-assistant-glow-28e9a8.yaml keeps referring to the online packages?

Or do I completely miss something and customizing the firmware is a lot easier than I think it is?

**Comment by phlexss** [comment, marked as answer]
This worked for me on Windows:

https://community.home-assistant.io/t/compile-esphome-firmware-updates-on-a-windows-computer/675385

- download glow package from github
- make the changes you want
- follow this guide to install python: https://esphome.io/guides/installing_esphome.html
- run "esphome config esp32.yaml" to check the config file (if you're using esp32, yaml is located in folder "home-assistant-glow")
- run "esphome compile esp32.yaml" to compile the package
- I then used the ESPhome Flasher Windows (https://github.com/esphome/esphome-flasher/releases) to flash the new firmware (home-assistant-glow\.esphome\build\home-assistant-glow\.pioenvs\home-assistant-glow\firmware.bin)
- configure wifi using https://web.esphome.io/

**Reply by FabsFabios** [reply]
Perfect! Thanks for the help @phlexss

**Reply by klaasnicolaas** [reply]
In my opinion this is a rather cumbersome solution, if all goes well you could also just copy the component YAML code and place it in your new YAML file (in ESPHome dashboard), this should then overrule the package import.

Another option is that I can change the package_import settings so that during an adoption the complete YAML code is fetched as a single file and placed in your ESPHome dashboard.

**Reply by phlexss** [reply]
I'm sorry, but I'm new to ESPhome on Home Assistant; how would I go about doing that?

You're saying it's possible to put the (changed) code from pulse_meter.yaml on Home Assistant so there's no need to re-compile and flash new firmware?

Where exactly would I need to put the code? In standard configuration.yaml or somewhere else? Do you have an example?

Thank you for your help!

**Reply by FabsFabios** [reply]
@klaasnicolaas  That would be awesome! If that's not possible I am wondering though, whether this could be added as a user changeable option in the same menu where the IMP/kWh can be changed?

Just like @phlexss though I am wondering how do I go about doing that? By "copy the component YAML code and place it in your new YAML file" you mean copy every code out of your three YAML files and combine it into my already existing home-assistant-glow-28e9a8.yaml file on ESPHome?

**Reply by klaasnicolaas** [reply]
> If that's not possible I am wondering though, whether this could be added as a user changeable option in the same menu where the IMP/kWh can be changed?

I don't quite understand what you mean. But no, this is not an changable option that you can use in the web server.

As an example, suppose you want to change something in the `pulse_meter` component, which is in a separate file from the repository and which is automatically imported during firmware compile.

You then copy the entire `pulse_meter` component from the import and place it at the bottom of the YAML config file (which is very compact) of your Home Assistant Glow in the ESPHome dashboard. You can then make the desired adjustments and then start the install sequence in your ESPHome dashboard to get the adjustments on your Home Assistant Glow.

**Reply by phlexss** [reply]
Hi @klaasnicolaas, thank you for all the info but so far I have no idea to what config file on Home Assistant you are referring.

Do you mean it's possible to make a change directly on HA somewhere (so there is no need to recompile) or are you talking about, for instance, adding the changes to the bottom of esp32.yaml before compiling?

....it would be great if it was possible to make these kind of changes without the need to recompile and flash. Is it possible?

It would be really be helpful if you can be a bit more specific about where/how to make these changes!

Thank you!

**Reply by klaasnicolaas** [reply]
If you add the Home Assistant Glow to ESPHome dashboard (device adoption), a separate YAML file is created. You can access this, for example, if you press the "edit" button. This is what I refer to all the time.

![image](https://github.com/klaasnicolaas/home-assistant-glow/assets/20448157/412394df-ca41-4b25-b36e-8933519cd245)

> are you talking about, for instance, adding the changes to the bottom of esp32.yaml before compiling?

No, not the esp32.yaml, the device YAML config in your ESPHome dashboard.

Changes you make in the YAML config of your ESPHome device must always be recompile and flashed again...

./Klaas

**Reply by phlexss** [reply]
I understand now. I didn't know about the adopt option in ESPhome (glow is my first esphome implementation).

Thanks for the info!

**Reply by phlexss** [reply]
One more question: I see now how I can edit the config from ESPhome dashboard.....I copied entire contents from pulse_meter.yaml to the config like you said. But when I "Install" I get errors:

Duplicate key "api"

If I remove the api part and install I get:

ID button_reset_total redefined! Check button->1->id.

How to fix this? How to make it overwrite the existing configuration because it's seeing the entities in the config already exist.

Thanks!

**Comment by klaasnicolaas** [comment]
FYI: I've created a PR (https://github.com/klaasnicolaas/home-assistant-glow/pull/420) which makes it easier to make adjustments in ESPHome, because from now on the entire config will be imported into your dashboard after adoption.

**Reply by phlexss** [reply]
Great!

**Reply by phlexss** [reply]
Is it correct that editing the yaml config and compiling from within HA doesn't work? ....because I get duplicate key errors when I recompile.....

**Reply by phlexss** [reply]
I haven't tried this in v4.01 yet....so I'll do that first....
